Istio troubleshooting guide on install fail due to mod resource (#813)
kyma-project/istio: docs/user/troubleshooting/03-70-reconciliation-fails-on-istio-install.md

> # Istio ingress-gateway Deployment, istiod Deployment or Istio CNI DaemonSets are not reconciling back to the desired state after being modified

Reviewer: ```suggestion
# Changes to Istio Ingress Gateway Deployment, Istiod Deployment, or Istio CNI DaemonSets Are Not Reverted After Reconciliation
```

## Symptoms

> - The Istio Ingress Gateway Deployment, Istiod Deployment, or istio-cni-node DaemonSets are not getting up

Reviewer: ```suggestion
- The Istio Ingress Gateway Deployment, Istiod Deployment, or  Istio CNI DaemonSets fails to start.
```

> - A non-default container is present in any of the above, and cannot start

Reviewer: ```suggestion
- Any of these resources contains a non-default container that is unable to start.
```

> - Istio Custom Resource is in Error state

Reviewer: ```suggestion
- The Istio custom resource is in the `Error` state.
```

> ## Causes

Reviewer: ```suggestion
## Cause
```

> - The Istio Ingress Gateway Deployment, Istiod Deployment, or istio-cni-node DaemonSets might have a container injected by an outside component, for example with a mutating webhook. If the new container cannot start for any reason, the pod will not get to a Running state, failing Istio module reconciliation.

Reviewer: ```suggestion
- If an external component, like a mutating webhook, adds a container to the Istio Ingress Gateway Deployment, Istiod Deployment, or Istio CNI DaemonSets, and that new container fails to start for any reason, the Pod is unable to reach the `Running` state. As a result, the Istio module reconciliation fails.
```

## Remedy

> 1. Check if a given resource's pod template is modified with additional container

Reviewer: ```suggestion
1. Check if a given resource's Pod template has been modified to include an additional container.
```

Reviewer: Don't you want to add commands with placeholders to these steps so the user can just run them? If possible, of course. Now, we tell what to do but not how to do it

> 2. Check if the new injected container is not getting up in the pod

Reviewer: ```suggestion
2. Check if the newly created container fails to start.
```

> 3. Remove new additional injected container from the pod template in the given resource

Reviewer: ```suggestion
3. If the container is unable to start, remove it from the Pod template of the given resource.
```

> 4. If error persists, check if there might be a mutating webhook in the cluster that is modifying Istio resources

Reviewer: ```suggestion
4. If the error persists, check if there is a mutating webhook in the cluster that is modifying Istio resources.
```
Maybe a silly question, but what should the user do if it turns out there is a webhook?

# Istio ingress-gateway Deployment, istiod Deployment or Istio CNI DaemonSets are not reconciling back to the desired state after being modified

## Symptoms

- The Istio Ingress Gateway Deployment, Istiod Deployment, or istio-cni-node DaemonSets are not getting up

> - Something new is stuck in the pod, and does not get up

Reviewer 2: ```suggestion
- A non-default container is present in any of the above, and cannot start
```

> - Istio Module Operator return error on Istio Install phase

Reviewer 2: ```suggestion
- Istio Custom Resource is in Error state
```

## Causes

> - The Istio Ingress Gateway Deployment, Istiod Deployment, or istio-cni-node DaemonSets pods template is modified, the new container is injected. The new container cannot get up in the pod for some reason, so the pod does not have status Running, and the install phase of the reconciliation fails.

Reviewer 2: ```suggestion
- The Istio Ingress Gateway Deployment, Istiod Deployment, or istio-cni-node DaemonSets might have a container injected by an outside component, for example with a mutating webhook. If the new container cannot start for any reason, the pod will not get to a Running state, failing Istio module reconciliation.
```

## Remedy

1. Check if a given resource's pod template is modified with additional container
2. Check if the new injected container is not getting up in the pod

> 3. Remove new additional injected container from the pod template in the given resource

Reviewer 2: This might not work
If the injection is controlled by a mutating webhook, simple remove will not fix the issue, as this needs to be configured in the component that injects the container

Author: Good point, I'll add a generic step to check for any mutating webhooks that might cause the problem

# Changes to Istio Ingress Gateway Deployment, Istiod Deployment, or Istio CNI DaemonSets Are Not Reverted After Reconciliation

## Symptoms

- The Istio Ingress Gateway Deployment, Istiod Deployment, or  Istio CNI DaemonSets fails to start.
- Any of these resources contains a non-default container that is unable to start.
- The Istio custom resource is in the `Error` state.

## Cause

- If an external component, like a mutating webhook, adds a container to the Istio Ingress Gateway Deployment, Istiod Deployment, or Istio CNI DaemonSets, and that new container fails to start for any reason, the Pod is unable to reach the `Running` state. As a result, the Istio module reconciliation fails.

## Remedy

1. Check if a given resource's Pod template has been modified to include an additional container.
2. Check if the newly injected container fails to start.

> [!TIP]

> > To retrieve information for the above steps, you can use command `kubectl describe <resource> -n istio-system -o yaml`

Reviewer: ```suggestion
> To retrieve information for the above steps, you can use the command `kubectl describe {RESOURCE_NAME} -n istio-system -o yaml`.
```

3. If the container is unable to start, remove it from the Pod template of the given resource.

>    `kubectl edit <resource> -n istio-system`

Reviewer: ```suggestion
   `kubectl edit {RESOURCE_NAME} -n istio-system`
```
